From d366928f36d6eeacd3417790148028f4c721944b Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:11:01 -0400
Subject: [usb] pl2303: switch to generic TIOCMIWAIT implementation

Message-id: <1410887463-20674-590-git-send-email-dzickus@redhat.com>
Patchwork-id: 95405
O-Subject: [RHEL7 PATCH 589/591] USB: pl2303: switch to generic TIOCMIWAIT implementation
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit a4bcb2945926901bbacf978eb0c0fdda5b327dc7
Author: Johan Hovold <jhovold@gmail.com>
Date:   Thu Jan 2 22:49:23 2014 +0100

    USB: pl2303: switch to generic TIOCMIWAIT implementation

    Switch to the generic TIOCMIWAIT implementation which does not suffer
    from the races involved when using the deprecated sleep_on functions.

    Acked-by: Arnd Bergmann <arnd@arndb.de>
    Signed-off-by: Johan Hovold <jhovold@gmail.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/serial/pl2303.c b/drivers/usb/serial/pl2303.c
index acdaae4..96c7240 100644
--- a/drivers/usb/serial/pl2303.c
+++ b/drivers/usb/serial/pl2303.c
@@ -716,46 +716,6 @@ static int pl2303_carrier_raised(struct usb_serial_port *port)
  return 0;
 }
 
-static int pl2303_tiocmiwait(struct tty_struct *tty, unsigned long arg)
-{
- struct usb_serial_port *port = tty->driver_data;
- struct pl2303_private *priv = usb_get_serial_port_data(port);
- unsigned long flags;
- unsigned int prevstatus;
- unsigned int status;
- unsigned int changed;
-
- spin_lock_irqsave(&priv->lock, flags);
- prevstatus = priv->line_status;
- spin_unlock_irqrestore(&priv->lock, flags);
-
- while (1) {
-  interruptible_sleep_on(&port->port.delta_msr_wait);
-  /* see if a signal did it */
-  if (signal_pending(current))
-   return -ERESTARTSYS;
-
-  if (port->serial->disconnected)
-   return -EIO;
-
-  spin_lock_irqsave(&priv->lock, flags);
-  status = priv->line_status;
-  spin_unlock_irqrestore(&priv->lock, flags);
-
-  changed = prevstatus ^ status;
-
-  if (((arg & TIOCM_RNG) && (changed & UART_RING)) ||
-      ((arg & TIOCM_DSR) && (changed & UART_DSR)) ||
-      ((arg & TIOCM_CD)  && (changed & UART_DCD)) ||
-      ((arg & TIOCM_CTS) && (changed & UART_CTS))) {
-   return 0;
-  }
-  prevstatus = status;
- }
- /* NOTREACHED */
- return 0;
-}
-
 static int pl2303_ioctl(struct tty_struct *tty,
    unsigned int cmd, unsigned long arg)
 {
@@ -834,7 +794,14 @@ static void pl2303_update_line_status(struct usb_serial_port *port,
   usb_serial_handle_break(port);
 
  if (delta & UART_STATE_MSR_MASK) {
+  if (delta & UART_CTS)
+   port->icount.cts++;
+  if (delta & UART_DSR)
+   port->icount.dsr++;
+  if (delta & UART_RING)
+   port->icount.rng++;
   if (delta & UART_DCD) {
+   port->icount.dcd++;
    tty = tty_port_tty_get(&port->port);
    if (tty) {
     usb_serial_handle_dcd_change(port, tty,
@@ -954,7 +921,7 @@ static struct usb_serial_driver pl2303_device = {
  .set_termios =  pl2303_set_termios,
  .tiocmget =  pl2303_tiocmget,
  .tiocmset =  pl2303_tiocmset,
- .tiocmiwait =  pl2303_tiocmiwait,
+ .tiocmiwait =  usb_serial_generic_tiocmiwait,
  .process_read_urb = pl2303_process_read_urb,
  .read_int_callback = pl2303_read_int_callback,
  .probe =  pl2303_probe,
-- 
1.7.1